assets/app/view/part/town_rect.rb

```diff
@@ -11,111 +12,171 @@ class TownRect < Base
+            [nil, :sharp, :gentle, :straight][(edge_a - edge_b).abs]
+          end
+      end
+
+      # Returns a symbol of :left, :right, or :straight based on the direction
+      # the track curves starting from edge_a
+      def track_direction
+        @track_direction ||=
           begin
             edge_a, edge_b = normalized_edges
 
-            case edge_b - edge_a
-            when 3
+            if (edge_a - edge_b).abs == 3
               :straight
-            when 2
-              :gentle
-            else
-              :sharp
+            elsif edge_a > edge_b
+              :right
+            elsif edge_a < edge_b
+              :left
             end
           end
       end
 
       # Returns an array of options for track positions
       def track_location
-        edge_a, edge_b = normalized_edges
+        edge_a, = normalized_edges
 
-        case track_type
-        when :sharp
-          [SHARP_TRACK_LOCATIONS[edge_a % 6]]
-        else
-          [CENTER,
-           EDGE_TRACK_LOCATIONS[edge_a % 6],
-           EDGE_TRACK_LOCATIONS[edge_b % 6]]
-        end
+        loc =
+          if @tile.towns.size == 1
+            SINGLE_TOWN_REGIONS[track_type][min_edge]
+          else
+            EDGE_TOWN_REGIONS[track_type][edge_a % 6]
+          end
+        [loc]
       end
 
       # Returns an array of rotation options for the town rectangle that
       # corresponds to the positions given from track_location
       def rotation_angles
         edge_a, = normalized_edges
-        case track_type
-        when :sharp
-          [(edge_a + 2) * 60]
-        when :gentle
-          [(edge_a * 60) - 30,
-           (edge_a * 60) - 10,
-           (edge_a * 60) - 50]
+
+        if @tile.towns.size == 1
+          {
+            straight: [min_edge * 60],
+            sharp: [(min_edge + 2) * 60],
+            gentle: [(min_edge * 60) - 30],
+          }[track_type]
         else
-          [edge_a * 60] * 3
+
+          tilt = RECTANGLE_TILT[track_type] || 0
+
+          delta = {
+            left: -tilt,
+            right: tilt,
+            straight: 0,
+          }[track_direction]
+
+          [(edge_a * 60) + delta]
         end
       end
 
       # Returns an array of weights, location and rotations
       def position
-        edge_a, edge_b = normalized_edges
-        angles = case track_type
-                 when :sharp
-                   [(edge_a + 0.5) * 60]
-                 when :gentle
-                   [(edge_a + 1) * 60,
-                    (edge_a * 60) + 2,
-                    (edge_b * 60) - 2]
-                 else
-                   [edge_a * 60] * 3
-                 end
-        radians = angles.map { |a| a / 180 * Math::PI }
+        edge_a, = normalized_edges
+
+        if @tile.towns.size == 1
+          angles = {
+            straight: [edge_a * 60],
+            sharp: [(min_edge + 0.5) * 60],
+            gentle: [(min_edge + 1) * 60],
```

Comment:
probably better to do case statments here so you're not allocation dictionaries

Reply:
9488efb